ログイン用のアクションを追加

a5ob7r opened the pull request:
# 概要

ユーザのフォーム入力を経てAPIサーバに送信し、アクセストークンを発行するアクションを追加する。

- 発行したアクセストークンはRDBをバックエンドとしたセッションストアに保存する。
- ファイル一覧表示用のアクションも同時に実装しているが、これはログイン済みの場合のリダイレクト先として利用しているためである。また、本PRで必要な範囲の機能のみを実装しており、メインのロジックやテストは、別PRで扱う予定である。

# 補足

本PRは #6 の後続である。

a5ob7r committed:
Add the empty action as the login action

a5ob7r committed:
Add the file index action as an empty action

a5ob7r committed:
Add the login action

a5ob7r committed:
Add some logic to the files index action

a5ob7r commented on spec/system/session_creation_spec.rb:
```diff
@@ -0,0 +1,167 @@
+        click_button 'Log in'
+      end
+
+      it 'ログイン用のURLにリダイレクトする' do
```
ログアウトした旨をエラーメッセージとして表示しようかと思ったけど、ログイン画面が表示されているということは、ログインする必要があることがわかるし、それはそのまま現在ログアウト状態であることを示すので、わざわざ表示しなくてもいいかな。

KT-tim replied:
まあここは適当でいいと思います(案件のシステムでも出して無いです)

tk0miya replied:
いいと思う

a5ob7r commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,37 @@
+    params.require(:access_token).permit :email, :password
+  end
+
+  def logged_in_user
```
このメソッド名はあんまりしっくりこないなー。というか、#before_actionで利用するメソッドの命名はどう考えたらいいのかな。

tk0miya replied:
僕は require_, validate_, ensure_, check_ あたりを好みますね。

契約プログラミングにおける事前条件に近いものとして使うのが好きなので「期待する状態であること」を確認したり検証したりする、という位置づけで命名するのが好みです。
https://ja.wikipedia.org/wiki/%E5%A5%91%E7%B4%84%E3%83%97%E3%83%AD%E3%82%B0%E3%83%A9%E3%83%9F%E3%83%B3%E3%82%B0

a5ob7r replied:
それに従うと、今回なら、 `check_logged_in_user` とか `check_user_logged_in` とかになるのかな。でも、メソッド名からは場合によってはリダイレクトするということが読み取れないんだよな。

a5ob7r replied:
このフィルターの意図は、まだログインしていないユーザのリクエストのみを受け付けるということだから、 `require_not_logged_in` とかならいいのかな。前述のリダイレクトすることが読み取れない問題は解決していないけど、ログインしているのならエラーを返すとかリダイレクトするなどして、とにかくログイン済みのユーザを弾くのだなということは読み取れるそうだ。

tk0miya replied:
コメント不足でした。check_ は表現としてあいまいなので、なるべく他の具体的な動詞を使うことを推奨します。チェックするというのは何かを説明しているようで、具体性に欠けることもあって、require_, validate_ といった具体的な表現のほうが良いですね。

ですから、 `require_not_logged_in` (未ログインであることを期待する/求める) というのはよい表現だと思います。

Taku3939 commented:
マージ先をmainにするとfaker gemの追加などの他のPRの変更が混ざってしまっているので変更箇所がわかりにくいように感じました。
- draftにして#6 のPRがマージされるのを待つ
- #6からブランチを切ってマージ先を#6のブランチにする
の方が良いように感じたのですがどうでしょうか？

tk0miya commented:
今週は月曜が休みだったお陰で火曜が忙しかった & 体調を崩した人がいたこともあって、レビューが長引いているものの、さくさく #5 をマージするのが理想ですね。
ブランチの向き先を変えると差分は読みやすくなるのだけど、マージしてはいけない状態になるので扱いづらいし、レビューを頑張る、マージしやすい PR にする、という戦略がいいと思う。

※ その点でいうと、#6 については session_store の変更だけっぽいので、#5 に依存しない形式にするのは手ですね

Taku3939 commented on spec/system/session_creation_spec.rb:
```diff
@@ -0,0 +1,167 @@
+          status:  200,
+          headers: { 'Content-Type' => 'application/json' }
+        )
+        WebMock.stub_request(:get, "#{Api::User.base_url}/files").to_return(
```
~URI.joinでも書けそう~
https://docs.ruby-lang.org/ja/latest/method/URI/s/join.html

tk0miya replied:
参考まで: https://qiita.com/ryonext/items/0bfd2592d713211bbc2f

a5ob7r replied:
むしろ、最初は `File.join` を使って書いていたんですけど、string interpolationを使った書き方の方が見やすいなと思って変更したんですよね。でも、たとえば今回の `Api::User.base_url` の末尾に `/` がついているかどうかで結果が変わるから、やっぱり元に戻したほうがいいか。結合対象が全部リテラルの場合でもなければ、 `File.join` を使うのが吉かな。

a5ob7r replied:
そういえば、 `File.join` はOSによって結合に利用されるパスセパレータが異なるので、ファイルパスの結合以外の用途で使うべきではない、みたいな話はないのかな。

まあ、アプリケーションであれば実行されるプラットフォームは大体決まっているので、心配するようなことではないとは思う。

epaew replied:
`File.join` vs `URI.join` をやりがちだけど、

- 基本的には URL の組み立てには `URI.join` を使うべきじゃねえの
- path 部分の組み立てには `File.join` や `Pathname.join` を使うのがいいんじゃないの

と思ったりする。

```ruby
URI.join("http://example.com", File.join(...))
```

「ただ URL を組み立てるだけで大げさｗ」 とも思わなくはないけれど。  
「どっちを使うか」以外の選択肢もあるんじゃないのとは思う。

KT-tim commented on app/views/sessions/new.html.erb:
```diff
@@ -0,0 +1,11 @@
+<%= form_for :access_token, url: sessions_path do |f| %>
```
今はform_withを使うのが主流なので、特にこだわりなければwithをおすすめします（もしかして他の箇所でもform_with使ってたっけ？見逃してるかも）

a5ob7r replied:
そうなんですね。知りませんでした。

https://railsguides.jp/5_1_release_notes.html#form-forとform-tagのform-withへの統合

Railsガイドでも `#form_with` を使ってフォームを組み立てているし、なんならsoft-deprecatedだと書いてありますね。

> Before form_with was introduced in Rails 5.1 its functionality used to be split between [form_tag](https://api.rubyonrails.org/v5.2/classes/ActionView/Helpers/FormTagHelper.html#method-i-form_tag) and [form_for](https://api.rubyonrails.org/v5.2/classes/ActionView/Helpers/FormHelper.html#method-i-form_for). Both are now soft-deprecated. Documentation on their usage can be found in [older versions of this guide](https://guides.rubyonrails.org/v5.2/form_helpers.html).

https://guides.rubyonrails.org/form_helpers.html#dealing-with-basic-forms
https://guides.rubyonrails.org/form_helpers.html#using-form-tag-and-form-for

a5ob7r committed:
Use #form_with instead of #form_for

The former is already soft-deprecated.

https://guides.rubyonrails.org/form_helpers.html#using-form-tag-and-form-for

a5ob7r committed:
Change a before_action's method name

This is probably more understandable for the code intention.

a5ob7r committed:
Add a concern as session utilities

We no need to concern any access tokens in sessions directly when handle
sessions.

tk0miya commented on app/controllers/application_controller.rb:
```diff
@@ -7,4 +7,11 @@ def render_internal_server_error
+
+  # #reset_session と異なりセッションIDのみをリセットする。
+  def reset_session_id
+    except_session_id = session.to_h
```
これって session_id を取り除いたりしているんでしたっけ。

コードを見た限りだと `except` 感がよく分からなかったです。
読んだ印象だと `original_session` あたりのネーミングが妥当そうに見えました。

a5ob7r replied:
`session.to_h` の返り値にsession_idは含まれないです。ですが、そのことはこの呼び出しからはわからないので、そのことをコメントに書こうと思ったのですが、それなら変数名で表現したほうがいいのでは、と考えてこのような変数名にしました。

セッションストアに保存されている値のうち、セッションID以外をセッションの値と解釈するのが一般的ということであれば、original_sessionとかold_sessionとかでもいいかと思います。つまりは、`session.to_h` の返り値にsession_idは含まれないということが多くの人の共通認識であるなら、別の変数名もありだと考えています。

tk0miya replied:
Rails が提供する session は `#session` でアクセスできる魔法の Hash というものが共通認識で、session_id がなにかというのは意識せずに済むようになっているはずです。

言い換えると、意識せずに済んでいた session_id を意識せざるを得ないようにしているのだけど、コード上には session_id が登場しないので、何をやっているのかが分かりづらいコードに見えています。

a5ob7r replied:
なるほど。そういうことであれば、先の変数名は変更したほうがよさそうですね。

a5ob7r replied:
session_idに対する認識については承知しましたが、後の変更でこの関数自体が削除されたので、この話題は解決したということにしたいと思います。

tk0miya commented on app/controllers/concerns/sessions.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module Sessions
```
認証まわりのメソッドが集まっている concern なので `Authenticatable` あたりの名前が良さそうに思いました。

* 複数形にしている理由がよく分からなかった
* セッション == 認証ではないので、中身とモジュール名が不一致だと感じた

a5ob7r replied:
認証、なんですかね。正直、適当な名称が思いつかなかったので、とりあえずつけた名称でした。

sessionは手段で、実態は提供されるメールアドレスとパスワードを元にユーザを識別および認証しているのか。

tk0miya replied:
login/logout というのは認証にまつわる動作とするのが一般的ですね。

API レベルでは

* 最初のアクセスでは「予め登録したメールアドレスとパスワードの組を知っている人」をユーザーであると認証している
* その後は「払い出したアクセストークンを持っている」クライアントをあるユーザーであると認証する

というセッション管理をしていますよね。

一方、こちらのクライアント側では、Rails のセッションの仕組みを使って

* 最初のアクセスでは「予め登録したメールアドレスとパスワードの組を知っている人」をユーザーであると認証し、セッションにアクセストークンを保存している
* その後は「セッションにアクセストークンが保存されている」ブラウザをユーザーであると認証する

というセッション管理をしていますよね。

>sessionは手段で、実態は提供されるメールアドレスとパスワードを元にユーザを識別および認証しているのか。

クライアント側では session にトークンをセットすることを login, session にトークンが入っているかのチェックを logged_in? で扱っていますが、これは立派な認証だと思いますよ。

今日話しをしたとおり、セッション(データ) というのは一連のやり取りにおけるデータの保存場所、という意味しかもっていません。
単なるデータ置き場に「ログインする」「ログイン済みか判定する」という機能を付与しているのは、認証のためですよね?

tk0miya commented on app/controllers/files_controller.rb:
```diff
@@ -0,0 +1,18 @@
+
+  def index
+    # TODO: Implement the main logic.
+    reset_session_id
```
ここで session_id をリセットしているのはなぜなんでしたっけ。

a5ob7r replied:
不要だったので削除しました。

a5ob7r committed:
Use File.join to join URLs and paths

This is a more robust way to join them about path separators.

```ruby
uri1 = 'http://foo.com'
uri2 = 'http://foo.com/'

"#{uri1}/bar" # => 'http://foo.com/bar'
"#{uri2}/bar" # => 'http://foo.com//bar'

File.join uri1, 'bar' # => 'http://foo.com/bar'
File.join uri2, 'bar' # => 'http://foo.com/bar'
``

a5ob7r commented:
修正が必要な箇所については対応が完了しました。

tk0miya commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,6 +1,39 @@
+
+  def create
+    response = Api::AccessToken.create access_token_params
+                                         .transform_keys { |k| k == :email ? :username : k }
```
ここで transform をするのであれば、最初から :username で送った方がよいのでは?
敢えて一度 :email で受け取っているのはなんのためでしょうか

a5ob7r replied:
前提として、ユーザに入力してもらいたいデータはユーザ登録時に利用したメールアドレスなので、ラベルのテキストは `Email` にすることを考えています。

> ここで transform をするのであれば、最初から :username で送った方がよいのでは?
> 敢えて一度 :email で受け取っているのはなんのためでしょうか

これはつまり、コントローラではなくビューで `:email` と `:username` の読み替えをしたらどうか、ということですよね。

現行の仕様について、何かこだわりがあるわけではないです。強いていうなら、ユーザにはメールアドレスを入力させるのであり、それを `username` パラメータとして送信したいのはクライアントサーバ(とAPIサーバ)の都合なので、ユーザに直接接するビューよりも、よりクライアントサーバの内部寄りのコントローラで読み替えるほうが自然に感じた、くらいでしょうか。

正直どちらでもいいとは思います。そして、どちらが良いかを判断する基準がわからないので、個人の好みで現状の実装を選びました。

a5ob7r replied:
基本的にコントローラにロジックは書かないほうがよいとすると、ビューで読み替えるほうがいいのかな。

tk0miya replied:
ラベル名を Email にすることと、入力フォームのデータ名を email にするのは別物ですよね。
ラベル名を「メールアドレス」に変更してほしいという連絡が来た場合、どうしますか?

名前が揃っていたほうが気持ち良いというのはわかります。
一方で、どこかで読み替えをする必要がある今回のようなケースで、email というデータ名にする価値がどのくらいあるのか、というのがポイントになりそうです。
このシステム (クライアント側) にとって email という名前に価値があるのであれば、フォームでは email としておいて、API 呼び出し時に名前を読み替える意味は大いにありそうです。
一方で、今作っているこの規模の範囲でいうと「このシステムにとって」そのデータをどう呼ぶのかはあまり重要ではないように見えます。

少なくとも現状のコードでは敢えて email と名付けたあとに、わざわざ読み替えを行っているように見えます。

※ チュートリアルとして実装している、実験用コードなので意地悪なものいいかも知れませんね。悪意はありません。

tk0miya replied:
もう少しシンプルに言うと、そのデータを `email` と呼びたい理由がラベル名以外のあるのかが気になっています。
理由があるのであれば、コントローラで変換するのに異論はありません。

a5ob7r replied:
> もう少しシンプルに言うと、そのデータを `email` と呼びたい理由がラベル名以外のあるのかが気になっています。
> 理由があるのであれば、コントローラで変換するのに異論はありません。

おそらく根底にあるのは、「メールアドレスをusernameとして扱うのはAPIサーバの都合でしかなく、やはりメールアドレスはemailという変数名で扱いたい、なぜならusernameという変数名からはその値がメールアドレスであるとは連想しづらいから」という発想だと思います。

同じことかもしれませんが、ユーザにEmailというラベルを表示するという前提があるとして、どこかで読み替えをする必要があるわけですが、それはバックエンド側の都合なので、それをわざわざフロントエンド、というかユーザインターフェース側であるビューでやる必然性があるのだろうか、と感じました。まあ、今回に関してはフォームのラベル自体は変化せず、タグのIDだけが変化するので、ユーザインターフェース側といっても、実のところバックエンド側と解釈してもいいのでは、とも考えています。

あとは、API側の仕様とはいえ、やはりusernameという名前からはメールアドレスは連想されないので、後でコードを読み直したときに、「あれ、これ間違いじゃないのか？」と考えそうだなと感じていることも理由の一つかもしれません。API呼び出しの直前で読み替えていると、なんとなくその読み替えはAPI呼び出しに必要なことなのだなと納得できそうですし、そこ以外の箇所ではemailという名称なら、先の勘違いも発生しないだろうと考えます。こちらについても、最初の読み替えのところでなぜ読み替えをしているかをコメントで残せば解決できるかもしれませんね。

ここまで書いていて、アプリケーション内において変数が持つ意味に注目するか、変数の値そのものの実態に注目しているか、という違いなのかなと感じました。要は抽象化して考えているか否かということかな。

tk0miya replied:
>同じことかもしれませんが、ユーザにEmailというラベルを表示するという前提があるとして、どこかで読み替えをする必要があるわけですが、それはバックエンド側の都合なので、それをわざわざフロントエンド、というかユーザインターフェース側であるビューでやる必然性があるのだろうか、と感じました。

ここで前回の問いにたどり着くわけです。

>ラベル名を Email にすることと、入力フォームのデータ名を email にするのは別物ですよね。
>ラベル名を「メールアドレス」に変更してほしいという連絡が来た場合、どうしますか?

ラベル名を「メールアドレス」に変更した場合、フォームのデータ名は `mail_address` に変更しますか?

アプリケーションというものは徐々に変化を遂げていくもので、もちろん UI というのは変化していきます。
当初は「Email」と表現していたものが「Eメール」になり、「メールアドレス」になり、さらに別のものになる可能性は十分にあります。
こうしたときにモデル、ビュー、コントローラーではどう扱うのかを考えたいですね。

※ 適当にググってみましたが、たとえば Hulu では「Eメールアドレス（ログインID）」と呼んでますね。ここらへん、システムによって同じメアドに対していくつかの呼び方がありそうです。

データの本質が変わっていないので内部名は `email` のままにするのか、ユーザー向けの表示に合わせて内部名も変更するのか。いずれかのコンポーネントのみ変更するのか。

僕は全社の「データの本質が変わっていないので〜」をイメージしており、その上で「今回のシステムは API を呼ぶためだけのものなので、データの本質は API に従って username」だと思ったのです。


> おそらく根底にあるのは、「メールアドレスをusernameとして扱うのはAPIサーバの都合でしかなく、やはりメールアドレスはemailという変数名で扱いたい、なぜならusernameという変数名からはその値がメールアドレスであるとは連想しづらいから」という発想だと思います。

僕は「今回のシステムは API サーバを使うためのシステム」として捉えているので、ここは API サーバに従っておいたほうが良いだろう、と考えました。ここで個性を出すことによって得する部分が思いつかなかったと言えるかも知れません。

tk0miya commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,6 +1,39 @@
+    flash[:success] = 'You logged in successfully'
+    redirect_to files_path, status: :see_other
+  rescue Flexirest::HTTPClientException => e
+    if e.result.error == 'invalid_request'
```
雑に聞いてしまうのだけど

* エラーを見分けるのは error 値でしか出来ないのでしょうか。この文字列を元に判定するのは脆弱じゃないのか気になりました
* invalid_request 以外のエラーって何がありましたっけ? 起こりうるエラーなのか、そうでないのかが気になりました。

について教えてもらえますか。

a5ob7r replied:
> * エラーを見分けるのは error 値でしか出来ないのでしょうか。この文字列を元に判定するのは脆弱じゃないのか気になりました

APIサーバの仕様書によると、クライアントエラーに関しては、ここではステータスコードは400の場合しかなく、エラーの識別はerror値で行うしかないです。これはOAuth 2.0の仕様だったと思います。正確には、OAuth 2.0の仕様上は401を返す場合もありますが、今回のAPIサーバでは401を返すべき状況は発生しません。

> * invalid_request 以外のエラーって何がありましたっけ? 起こりうるエラーなのか、そうでないのかが気になりました。

APIサーバの仕様書によると以下の4つを返し得るとありました。

- invalid_request
    - 必須パラメータの欠落
- invalid_grant
    - grant(今回であればメールアドレスとパスワード)が不正
- unsupported_grant_type
    - リクエストされたgrantタイプをサポートしていない
- invalid_scope
    - リクエストされたスコープが不正

このうち、今回のリクエストで発生し得る、というか、ユーザが対応可能なのは、invalid_grantのみです。フォームを利用してリクエストすれば、必須のパラメータは必ず含まれるのでinvalid_requestは発生しない。リクエストに含めるgrant_typeパラメータとscopeパラメータはクライアントサーバ側で設定するのでunsupported_grant_typeとinvalid_scopeは発生しない。もちろん、フォームやアクションが正しく実装されている前提ですが、仮にinvalid_grant以外のエラーが発生しても、それらはユーザからの入力とは無関係なので、500として扱うのが適当だと考えます。

ちなみに、コードでは、invalid_grantとするべきところがinvalid_requestになっていますね。修正します。

tk0miya replied:
説明ありがとう。説明通りであればこの分岐は必要なさそうだと思ったのですが、どうでしょうか。
不具合でしか発生しないものに対して、分岐を書くのは労力の割には得るものがすくなさそうです。
不具合が合った場合、フォームが表示されずにクラッシュするという感じですよね。

a5ob7r replied:
> 不具合が合った場合、フォームが表示されずにクラッシュするという感じですよね。

うーん、どうなんだろう。SessionsController#create用のビューは用意していないので、何もしなければクラッシュするのかな。

a5ob7r replied:
> 説明ありがとう。説明通りであればこの分岐は必要なさそうだと思ったのですが、どうでしょうか。
> 不具合でしか発生しないものに対して、分岐を書くのは労力の割には得るものがすくなさそうです。

ここでいう分岐を分岐とはどこのことを指しているのでしょうか。 `if e.result.error == 'invalid_request'` のことでしょうか。もしそうなら、この分岐を書かないとすると、4XXが発生した際には、例えユーザの入力に原因がない場合でもフォームを再表示することになると思うのですが、この認識であってますかね。

tk0miya replied:
はい、その分岐を指しています。

上で回答してもらったけど、4XX が発生することはないはず (あったとして、サーバ/クライアントのバグ) ですよね。

>フォームを利用してリクエストすれば、必須のパラメータは必ず含まれるのでinvalid_requestは発生しない。リクエストに含めるgrant_typeパラメータとscopeパラメータはクライアントサーバ側で設定するのでunsupported_grant_typeとinvalid_scopeは発生しない。


* この変更をリリースする際はテストするはずなので、バグが発生する可能性は低いはず
* バグがあった場合に、500 エラーが出るのと、フォームが再表示されるのはユーザーにとってはあまり変わらないように見える
    * 500 エラーが出たからバグだろうと考えるのはプログラマだけです
    * 結構な割合で再試行することになると予想します
    * プログラマですら、(一時的な) インフラのトラブル? などと考えそうな気もします。

という思考の流れで、この分岐を作ってまで 500 エラーを出したいのか、と考えました。

a5ob7r replied:
> 上で回答してもらったけど、4XX が発生することはないはず (あったとして、サーバ/クライアントのバグ) ですよね。

いえ、ユーザがログイン時に入力するメールアドレスやパスワードを間違えれば、400が発生します。

> * バグがあった場合に、500 エラーが出るのと、フォームが再表示されるのはユーザーにとってはあまり変わらないように見える

しっくりはこないけど、想定されるフローを考えると、結局再試行することになりそうだから違いはないと考えていいのか。
   
>   * 500 エラーが出たからバグだろうと考えるのはプログラマだけです
>   * 結構な割合で再試行することになると予想します
>   * プログラマですら、(一時的な) インフラのトラブル? などと考えそうな気もします。
> 
> という思考の流れで、この分岐を作ってまで 500 エラーを出したいのか、と考えました。

なるほど。そう言われてみると、わざわざ500エラーを表示しても特にリターンはないのか。それに、どうせ再試行するという話もその通りですね。自分でもそうしそうです。となると、ビューではなくエラーメッセージの内容を分岐させるべきなのかな。

再試行することを考えると、5XXが発生しても、タイムアウトやコネクションの失敗の場合でも結局ユーザは再試行する気もするな。そうなると、それらが発生したとしても、わざわざ500を返す必要性はないということになるのだろうか。

tk0miya replied:
>>>もしそうなら、この分岐を書かないとすると、4XXが発生した際には、例えユーザの入力に原因がない場合でもフォームを再表示することになると思うのですが、この認識であってますかね。

という問いに対して

>>上で回答してもらったけど、4XX が発生することはないはず (あったとして、サーバ/クライアントのバグ) ですよね。

と回答しました。

>いえ、ユーザがログイン時に入力するメールアドレスやパスワードを間違えれば、400が発生します。

ですから、ユーザーの入力に原因があるものは除いて会話していたつもりです。
伝わりづらくてごめん


整理すると

* 4xx が発生するのは、以下の 4種類
    * invalid_request
    * invalid_grant
    * unsupported_grant_type
    * invalid_scope
* これらが発生する可能性は以下の通り
    * invalid_grant: ユーザーの入力間違いによって発生
    * その他: 起こり得ない。発生するとしたらサーバ/クライアントのバグによるもの

という状況という理解です。

こういう状況下において、分岐を用意してもお得ではないのでは、というのが僕の意見です。

API 定義上では起こりうるが、実装上起こり得ないエラーに対して備えておくべきか、という話だと思っています。


実際のシステムでは、ひとつのシステムの中で何十、何百も API を呼ぶ箇所があるので、コード量が 10倍, 100倍になっていきます。1箇所であれば大差がないコードであっても、100箇所になるとメンテにかかる時間が膨れ上がることになるので、これって本当に必要なんだっけ? 誰かにとって価値があるコードなんだっけ? と考える習慣があるといいな、と思っています。

tk0miya replied:
>なるほど。そう言われてみると、わざわざ500エラーを表示しても特にリターンはないのか。それに、どうせ再試行するという話もその通りですね。自分でもそうしそうです。となると、ビューではなくエラーメッセージの内容を分岐させるべきなのかな。
>
>再試行することを考えると、5XXが発生しても、タイムアウトやコネクションの失敗の場合でも結局ユーザは再試行する気もするな。そうなると、それらが発生したとしても、わざわざ500を返す必要性はないということになるのだろうか。

単に 500 エラーを表示してもユーザーにとってメリットがないのはそのとおりですね。
ユーザーには具体的なメッセージ、たとえば「時間を空けてからお試しください」などを表示する方が効果があると思います。


一方、今回のコードに対してメッセージの分岐を行うべきかどうかは、今回の議論と同じ話にたどり着きそうです。
invalid_grant 以外のエラーは「プログラムにバグがなければ起こり得ない」ものですよね?
バグを起こさないように rspec を書いたり、手動で動作確認をしたりするわけで、これらのチェックをすり抜けてバグが表面化するということを気にかけるべきでしょうか。

コードを全く信頼できないと考えるのであれば、こうしたチェックは強化する価値がありそうですが、その場合はステータスコードだけで良いのかも検討が必要ですね。JSON が壊れていた場合どうするのか、といったことも考えるなど、考える幅が広がりそうな気がします。

バグの発生可能性を考えて、エラーチェックを厳密に行うのか、緩めに行うのかを考えられると良さそうです。

※ その上で、今回の場合は rspec や手動テストで検証できることもあり、対応は不要だと考えています。
※ 別の言葉でいうと、どういうときに分岐があると嬉しいのか、それは起こりうるのかを考えたいですね。

tk0miya commented on app/models/api/files.rb:
```diff
@@ -0,0 +1,14 @@
+    get :all, '/files'
+
+    before_request do |name, request|
+      if name == :all
```
質問: このテクニックはどこで知りました? この書き方を知らないので、ポインタを知りたいです。

a5ob7r replied:
おそらくドキュメントを眺めていて見つけたのだと思います。

```ruby
  before_request do |name, request|
    if request.post? || name == :save
      id = request.post_params.delete(:id)
      request.get_params[:id] = id
    end
  end
```

https://github.com/flexirest/flexirest/blob/master/docs/using-callbacks.md

tk0miya replied:
説明ありがと。一つ賢くなった！

tk0miya commented on spec/system/session_creation_spec.rb:
```diff
@@ -0,0 +1,167 @@
+      context 'APIサーバからエラーレスポンスが返ってきたとき' do
+        before do
+          WebMock.stub_request(:post, File.join(Api::User.base_url, 'signin')).to_return status: 500
+          WebMock.stub_request(:get, File.join(Api::User.base_url, 'files')).to_return(
+            body:    '{}',
+            status:  200,
+            headers: { 'Content-Type' => 'application/json' }
+          )
+        end
```
こっちの stub って必要なんでしたっけ?

a5ob7r replied:
不要だったので削除しました。

a5ob7r committed:
Remove an unnecessary newline for a method call

tk0miya commented on spec/system/session_creation_spec.rb:
```diff
@@ -0,0 +1,167 @@
+
+    context 'セッションが無効であるとき' do
+      before do
+        WebMock.stub_request(:post, File.join(Api::User.base_url, 'signin')).to_return(
+          body:    {
+            access_token: 'some_token',
+            token_type:   'bearer'
+          }.to_json,
+          status:  200,
+          headers: { 'Content-Type' => 'application/json' }
+        )
```
ログイン済みのとき、これいるんでしたっけ?

a5ob7r replied:
ログイン済みの状況を作るために、事前にフォームをポストしてログインするので必要です。

a5ob7r committed:
Change a temp variable name for #reset_session_id

We no need to concern any session ids on Rails, so the fast that
the return value of `session.to_h` doesn't contain `session_id` is
obvious.

a5ob7r committed:
Change the module name for authentication

a5ob7r committed:
Remove #reset_session_id

It's generally enough just to reset session id after each of logins.

a5ob7r committed:
Remove the key name transformation from the controller

Use the transformed name as the default id name for the form from the
beginning.

a5ob7r commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,6 +1,39 @@
+  def create
+    response = Api::AccessToken.create access_token_params
+                                         .transform_keys { |k| k == :email ? :username : k }
+                                         .merge(grant_type: 'password', scope: 'READ WRITE')
```
考えてみたら、このロジックもコントローラではなくリクエスト用のクラスに移動できるな。

a5ob7r replied:
#mergeを使って設定していたパラメータを、 `defaults` オプションを使ってリクエスト用のクラス側で設定するように変更しました。コントローラ上でやらなくてもよいロジックを他に逃がせたので良さげだ。

https://github.com/flexirest/flexirest/blob/master/docs/default-parameters.md

a5ob7r committed:
Move the parameter completion logic to the request client class

a5ob7r committed:
Remove some unnecessary mocks

a5ob7r committed:
Add #with to some mocks to test each of headers and body

a5ob7r committed:
Fix the invalid Authorization header's value format

tk0miya commented on app/controllers/files_controller.rb:
```diff
@@ -0,0 +1,16 @@
+
+  def index
+    # TODO: Implement the main logic.
+    Api::Files.all({ access_token: current_access_token })
```
実験してはいないのだけど、たぶん Hash オブジェクトじゃなく、キーワード引数で渡しても動くんじゃないかなあ
```suggestion
    Api::Files.all(access_token: current_access_token)
```

a5ob7r replied:
むむ、確かに。

a5ob7r replied:
キーワード引数でも動きました。修正します。

a5ob7r committed:
Change the argument passing style to keyword argument

a5ob7r committed:
Fix the error value used to identify the error

a5ob7r committed:
Add the status code to identify the error

a5ob7r commented on spec/system/session_creation_spec.rb:
```diff
@@ -0,0 +1,224 @@
+      end
+    end
+
+    context 'フォームへの入力に不正があるとき' do
+      context '各フォームへの入力が空文字列であるとき' do
+        let(:email) { '' }
+        let(:password) { '' }
```
ここのテスト、別に入力が空だから不正な入力ではないよな。あくまで、作成されていないユーザのgrantを送信するからエラーになるはずだ。

a5ob7r replied:
不要なlet文を削除して、スペックの文言も適切なものに変更した。

a5ob7r committed:
Fix the incorrect spec for invalid ROPC grant

Empty form inputs are not essential to test it because all of inputs
that are invalid as the user's resource owner password credentials grant
are the targets for this spec.

tk0miya reviewed (approved):
コメントをひとつ残しましたが、対応必須のものではないので +1 します。
